primitives/genesis-builder/src/lib.rs

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! Substrate genesis config builder
+//!
+//! This Runtime API allows to construct `GenesisConfig`, in particular:
+//! - serialize the runtime default `GenesisConfig` struct into json format,
+//! - put the GenesisConfig struct into the storage. Internally this operation calls
+//!   `GenesisBuild::build` function
+//! for all runtime pallets, which is typically provided by pallet's author.
+//! - deserialize the GenesisConfig from given json blob and put GenesisConfig into the state
+//!   storage. Allows to build
+//! customized configuration.
+//!
+//! Providing externalities with empty storage and putting GenesisConfig into storage allows to
+//! catch and build the raw storage of `GenesisConfig` which is the foundation for genesis block.
+
+sp_api::decl_runtime_apis! {
+	/// API to interact with GenesisConfig for the runtime
+	pub trait GenesisBuilder {
+		/// Get the default `GenesisConfig` as a JSON blob.
+		///
+		/// This function instantiates the default `GenesisConfig` struct for the runtime and serializes it into
+		/// a JSON blob.
+		///
+		/// # Returns
+		///
+		/// A `Vec<u8>` containing the JSON representation of the default `GenesisConfig`.
+		fn get_default_as_json() -> sp_std::vec::Vec<u8>;
+
+		/// Build `GenesisConfig` from a JSON blob and store it in the storage.
+		///
+		/// This function deserializes the `GenesisConfig` from the given JSON blob and puts it into the storage.
+		/// If the provided JSON blob is incorrect or the deserialization fails, this method will panic.
+		/// It is recommended to log any errors encountered during the process.
+		///
+		/// # Arguments
+		///
+		/// * `json` - A `Vec<u8>` representing the JSON blob containing the `GenesisConfig`.
+		fn build_from_json(json: sp_std::vec::Vec<u8>);
+
+		/// Patch default `GenesisConfig` using given JSON patch and store it in the storage.
+		///
+		/// This function generates the `GenesisConfig` for the runtime by applying a provided JSON patch.
+		/// The patch modifies the default `GenesisConfig` allowing customization of the specific keys.
+		/// The resulting `GenesisConfig` is then deserialized from the patched JSON representation and
+		/// stored in the storage.
+		///
+		/// If the provided JSON patch is incorrect or the deserialization fails, this method will panic.
```

Comment:
Maybe here there is too much description about what the provided (the reference) implementation does.

Probably the sentences should instead express what are the **expectations** from an implementation.

That is, if there are some constraints or expected behaviors, maybe is preferable writing it down in a form like
"the patching process should modify" instead of "the patching process modifies".

---

If we also expect that the implementation should strictly follow the reference implementation provided with this https://github.com/paritytech/substrate/pull/14310 then we also have to write it down

Reply:
You are welcome to improve wording with suggestion. I did my best, if you feel that it shall be improved, I encourage you to do so.

Reply:
Maybe I was not 100% clear.

I didn't want to say that wording is incorrect or the docs are incorrect. 
I just wanted to highlight that maybe implementation details should not go with the trait defs but with the implementation itself? 

In the first place because we didn't know how the trait will be implemented by a potential 3rd party implementor.

Reply:
I see your point. On the other hand documentation is also for user of this runtime API. So all these details allows to understand user what is going to happen, and also provides some implementation details to the implementator.

I can imagine that this API will work with some opaque, abstract `Vec<u8>` metadata blobs (not JSON), and runtime implementator may choose to use `toml` format for this. From this perspective `JSON` is just an implementation detail…

The end user must know what is intended input, and must also know what is intended behavior. The documentation of trait must work for both API user and API provider.

Reply:
Small adjustmens in: https://github.com/paritytech/substrate/pull/14131/commits/254776103e172630efd200cee13aecbc2a032e70